Stock values4decimal places

lomamech opened the pull request:
- Display unit cost with 4 decimal places

closes #6951

lomamech committed:
feature(Display Stock value)
- Display unit cost with 4 decimal places

closes #6951

lomamech committed:
feature(Set 4 decimales)
- Set 4 place for inventory unit cost

closes #6951

lomamech committed:
Clean and sanitaze code

jmcameron commented on server/controllers/stock/reports/stock_entry_depot.receipt.handlebars:
```diff
@@ -55,7 +55,7 @@
     <thead>
       <tr>
-        <th>{{translate 'STOCK.CODE'}}</th>
+        <th>{{translate 'STOCK.CODE'}} BBB</th>
```
It looks like you still have some debugging code here.

lomamech replied:
THANKS

jmcameron commented on client/src/js/filters/currency.js:
```diff
@@ -17,11 +17,10 @@ CurrencyFilter.$inject = [
   var requireCurrencyDefinition = false;
 
-  function currencyFilter(amount, currencyId) {
+  function currencyFilter(amount, currencyId, decimalSetting) {
```
Please add an @param comment in the function description above for the new decimalSetting parameter.

jmcameron replied:
I suggest renaming 'decimalSetting' to 'numDecimals' which is more descriptive.

jmcameron reviewed (commented):
This looks good (see my review comments).  However, I noticed that this update was not done on several reports:  shipping document, shipping manifest, and stock movement report.   I suggest you go back through and review all the reports to make sure this was not missed anywhere else.

Also, it is not clear to me why we have fractionalSize and 'decimalSetting'.  Why do we need both?  Could this be simplified?

I tested this on a couple of reports and it seems to work fine.

I view this as an interim fix.  Please see Issue https://github.com/IMA-WorldHealth/bhima/issues/6130 for my thoughts on this.

mbayopanda reviewed (commented):
Chris, I think this feature must optional since it's not all organizations that will want to have 4 decimals, for example, a hospital with enterprise currency FC; In Fc, there is no need for decimals which will produce huge numbers to read when it is not necessary.

<img width="371" alt="image" src="https://user-images.githubusercontent.com/5445251/223567266-391c4caa-3437-4b5c-a57a-ad67fe6308ec.png">

lomamech committed:
Improve: changement and sanitaze code

lomamech committed:
improvement(Set Values 4 decimal places)
- Display only 4 decimals if and only if the last number of decimals has been entered and if the value is less than 0.01

- Fix error in lot update form, support for acquisition date parameter when updating batch-related items

jmcameron commented on client/src/js/filters/currency.js:
```diff
@@ -68,26 +68,37 @@ function CurrencyFilter(CurrencyFormat, Session, $translate) {
     if (angular.isObject(number)) return '';
 
-    var isNegative = number < 0;
+    let isNegative = number < 0;
```
A minor comment:  isNegative is not used until line 172.  Therefor it could be computed as a 'const' on that line.

lomamech replied:
@jmcameron This file is really very old and it must date from the first version of BHIMA
````
    const isNegative = number < 0;
````
I just made the change

jmcameron commented on client/src/js/filters/currency.js:
```diff
@@ -68,26 +68,37 @@ function CurrencyFilter(CurrencyFormat, Session, $translate) {
-        hasExponent = false,
-        parts = [];
+    const numStr = `${number} `;
+    let formatedText = '';
```
In English, this is spelled:  formattedText

jmcameron reviewed (commented):
I looked over the code and made some OPTIONAL suggestions.   I realize that the function that you edited is old-style and complicated.  So it makes sense not to rewrite it since I doubt we have tests for it that will assure us that a rewritten version works correctly.  So feel free to ignore the coding suggestions I made.

I tested it and it seems to work well.   I did notice one possible problem.  If unit prices are less than 0.0001 (eg 0.000023), then the current algorithm will report 0.0000 (zero).  However I doubt this will occur in reality since per-unit stock prices will probably always be more than that.

So I think this PR is ready to go.

lomamech committed:
improve(Client filters currency)
- Improve code
- Update name of variable

jmcameron commented on server/controllers/stock/reports/stock_adjustment.receipt.handlebars:
```diff
@@ -126,7 +126,7 @@
             <td class="text-right">{{new_quantity}}</td>
             <td class="text-right">{{difference}}</td>
-            <td class="text-right">{{currency unit_cost ../enterprise.currency_id}}</td>
+            <td class="text-right">{{currency unit_cost ../enterprise.currency_id 4}}</td>
```
Question: Do you need to insert the '4' here now that you're forcing 4 if the number is too small?

lomamech replied:
Indeed, you will also notice that the option to display four decimals has only been used in very specific cases in cases where one seeks to display the unit cost of an inventory I had not thought generalized this rule throughout the system unless deemed necessary to apply everywhere.

jmcameron replied:
Sounds good!

jmcameron commented on client/src/js/filters/currency.js:
```diff
@@ -68,91 +68,110 @@ function CurrencyFilter(CurrencyFormat, Session, $translate) {
-        hasExponent = false,
-        parts = [];
+    const numStr = `${number} `;
+    let formattedText = '';
```
👍

jmcameron commented on client/src/js/filters/currency.js:
```diff
@@ -68,91 +68,110 @@ function CurrencyFilter(CurrencyFormat, Session, $translate) {
-    if (number === 0) {
-      isNegative = false;
-    }
+    const isNegative = number < 0;
```
👍

jmcameron reviewed (commented):
The code looks good.  I added one question about the code you might want to consider.

I think this is ready to go if @mbayopanda is happy with it.
